contentcuration/contentcuration/db/models/manager.py

```diff
@@ -556,15 +546,15 @@ def _deep_copy(
         excluded_descendants,
         can_edit_source_channel,
     ):
-
-        nodes_to_copy = self._all_nodes_to_copy(node, excluded_descendants)
-
         nodes_by_parent = {}
 
-        for copy_node in nodes_to_copy:
-            if copy_node.parent_id not in nodes_by_parent:
-                nodes_by_parent[copy_node.parent_id] = []
-            nodes_by_parent[copy_node.parent_id].append(copy_node)
+        # lock mptt source tree with shared advisory lock
+        with self.lock_mptt(node.tree_id, shared_tree_ids=[node.tree_id]):
```

Comment:
Could we change this to `nodes_to_copy = list(self._all_nodes_to_copy(node, excluded_descendants))` and then we can move the for loop out of the lock?

Reply:
Sounds good